_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,598 @@
+description:
+  "Tracking code size for embedded software projects can prevent surprises,
+  maintain productivity, and give insight to developers looking to retroactively
+  reduce a firmware binary size."
+tag: [fw-code-size]
+author: tyler
+image: /img/tracking-code-size/tracking-code-size.png
+---
+
+```
+arm-none-eabi-ld: firmware.elf section `.text' will not fit in region `ROM'
+arm-none-eabi-ld:: region `ROM' overflowed by 16 bytes
+make: error: arm-none-eabi-ld returned 1 exit status
+make: *** [firmware.elf] Error 1
+```
+
+This is the terminal output of my nightmares. It frequently means a halt in
+productivity and sends engineers scrambling to save 100 bytes wherever they can,
+all while trying to meet the deadlines and requirements set forth by the Product
+team.
+
+Fortunately, this message doesn't always have to come as a surprise. Where I've
+worked in the past, we had the tools and infrastructure in place to know when we
+would run out of code space, as well as where it all went. These tools were
+immensely helpful when our team at Pebble was trying to
+[cram an 800kB firmware into a 512kB ROM device](http://web.archive.org/web/20160308073714/https://blog.getpebble.com/2015/12/09/3ontintin/).
+
+Below is an example of a type of dashboard we had. When we approached 5kB
+remaining in our firmware image, we knew what to do.
+
+![](/img/tracking-code-size/running-out-code-space.png)
+
+Hopefully this post provides the tools you need to have a dashboard like this
+too!
+
+<!-- excerpt start -->
+
+Code size information should be surfaced and tracked throughout the development
+of an embedded software project, especially if there is even a slight chance of
+running out in the future. In this post, we'll cover why this is the case, show
+how to track code size, and show how to calculate code size deltas on pull
+requests to help keep code size to a minimum.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Why Track Code Size?
+
+If you are a firmware engineer, you've likely run into code sizes issues. If
+not, it is likely because the firmware engineers before you fought for an MCU
+with more micro-flash space or because the company went overboard and bought the
+most expensive chip.
+
+A team that has extra space in the firmware image will find a way to fill it,
+whether it is extra debug logs, a full-featured command-line interface, or fancy
+animations or images. With this in mind, it would be wise to constantly track
+and measure the code size of a project during development to know when and why
+the code size grows. This would mean that each commit and a handful of build
+configurations should be tracked so an accurate picture can be painted.
+
+I've seen many projects start to have a print out like below after each build,
```

Comment:
~~I've seen many~~ Most projects simply print out size information after each build. It may look like this:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      153108 B         1 MB     14.60%
            SRAM:       44544 B       256 KB     16.99%
        IDT_LIST:          56 B         2 KB      2.73%
```

This is step in the right direction, but it does not give us any sense of the changes in code size over time. Instead, we must track this information from commit to commit. With that information, your team can: 
* ...